src/app/chip-zcl-zpro/command-encoder/encoder.c

```diff
@@ -16,98 +16,67 @@
  */
 
 #include "chip-zcl-zpro-codec.h"
+#include <assert.h>
 #include <stdio.h>
 #include <string.h>
 
-uint32_t encodeApsFrame(uint8_t * buffer, uint32_t buf_length, uint16_t profileID, uint16_t clusterId, uint8_t sourceEndpoint,
-                        uint8_t destinationEndpoint, EmberApsOption options, uint16_t groupId, uint8_t sequence, uint8_t radius)
+static uint16_t doEncodeApsFrame(uint8_t * buffer, uint32_t buf_length, uint16_t profileID, uint16_t clusterId,
+                                 uint8_t sourceEndpoint, uint8_t destinationEndpoint, EmberApsOption options, uint16_t groupId,
+                                 uint8_t sequence, uint8_t radius)
 {
-    uint32_t size = 0;
-    if (buffer == NULL || buf_length == 0)
-    {
-        return size;
-    }
     size_t nextOutByte = 0;
 
-    if (nextOutByte >= buf_length)
-    {
-        return size;
-    }
+#define TRY_WRITE(dataItem)                                                                                                        \
```

Comment:
Do we have a generic method for interfacing with buffers? Having proper methods for interfacing with buffers may be a better solution than a locally defined macro

Reply:
something that works on top of PacketBuffer would be welcome. something that has a "measure" form would be ideal.

Reply:
I've just added BufBound, which might help here

Reply:
We'd need to switch this file to C++ to use that.  Which might be fine as long as we keep the entrypoint APIs extern C....  I can rebase on top of your change once it lands as desired, and make that modification.

Reply:
ah, yeah.  I was thinking that we probably want to just make all these files C++ and fix any void* stuff

Reply:
Yeah, https://github.com/project-chip/connectedhomeip/issues/1629 tracks the `void*` bits.

Reply:
https://github.com/project-chip/connectedhomeip/issues/2073 now tracks using `BufBound` here.